--- linux-stage.orig/fs/ext4/fsync.c	2014-03-12 21:32:43.000000000 +0800
+++ linux-stage/fs/ext4/fsync.c	2014-03-12 21:34:55.000000000 +0800
@@ -247,11 +247,17 @@ int ext4_sync_file(struct file *file, lo
 	    !jbd2_trans_will_send_data_barrier(journal, commit_tid))
 		needs_barrier = true;
 	jbd2_log_start_commit(journal, commit_tid);
+
+	mutex_unlock(&inode->i_mutex);
 	ret = jbd2_log_wait_commit(journal, commit_tid);
 	if (needs_barrier)
 		blkdev_issue_flush(inode->i_sb->s_bdev, GFP_KERNEL, NULL);
+	goto out_nolock;
+
  out:
 	mutex_unlock(&inode->i_mutex);
+
+ out_nolock:
 	trace_ext4_sync_file_exit(inode, ret);
 	return ret;
 }
