Index: linux-stage/fs/ext4/xattr.c
===================================================================
--- linux-stage.orig/fs/ext4/xattr.c
+++ linux-stage/fs/ext4/xattr.c
@@ -1354,7 +1354,7 @@ ext4_xattr_set_handle(handle_t *handle, 
 	}
 	if (!error) {
 		ext4_xattr_update_super_block(handle, inode->i_sb);
-		if (!(flags & XATTR_NO_CTIME))
+		if (!IS_NOCMTIME(inode) && !(flags & XATTR_NO_CTIME))
 			inode->i_ctime = ext4_current_time(inode);
 		if (!value)
 			ext4_clear_inode_state(inode, EXT4_STATE_NO_EXPAND);
Index: linux-stage/fs/ext4/namei.c
===================================================================
--- linux-stage.orig/fs/ext4/namei.c
+++ linux-stage/fs/ext4/namei.c
@@ -1442,7 +1442,8 @@ static int add_dirent_to_buf(handle_t *h
 	 * happen is that the times are slightly out of date
 	 * and/or different from the directory change time.
 	 */
-	dir->i_mtime = dir->i_ctime = ext4_current_time(dir);
+	if (!IS_NOCMTIME(dir))
+		dir->i_mtime = dir->i_ctime = ext4_current_time(dir);
 	ext4_update_dx_flag(dir);
 	dir->i_version++;
 	ext4_mark_inode_dirty(handle, dir);
Index: linux-stage/fs/ext4/acl.c
===================================================================
--- linux-stage.orig/fs/ext4/acl.c
+++ linux-stage/fs/ext4/acl.c
@@ -204,7 +204,8 @@ ext4_set_acl(handle_t *handle, struct in
 				return error;
 			else {
 				inode->i_mode = mode;
-				inode->i_ctime = ext4_current_time(inode);
+				if (!IS_NOCMTIME(inode))
+					inode->i_ctime = ext4_current_time(inode);
 				ext4_mark_inode_dirty(handle, inode);
 				if (error == 0)
 					acl = NULL;
