Index: linux-stage/fs/ext4/ext4_extents.h
===================================================================
--- linux-stage.orig/fs/ext4/ext4_extents.h
+++ linux-stage/fs/ext4/ext4_extents.h
@@ -113,6 +113,7 @@ struct ext4_extent_header {
  * Truncate uses it to simulate recursive walking.
  */
 struct ext4_ext_path {
+	unsigned long			p_generation;
 	ext4_fsblk_t			p_block;
 	__u16				p_depth;
 	struct ext4_extent		*p_ext;
