Index: linux-2.6.32-el6-beta/fs/ext4/ext4_extents.h
===================================================================
--- linux-2.6.32-el6-beta.orig/fs/ext4/ext4_extents.h
+++ linux-2.6.32-el6-beta/fs/ext4/ext4_extents.h
@@ -201,6 +201,11 @@ static inline unsigned short ext_depth(s
 	return le16_to_cpu(ext_inode_hdr(inode)->eh_depth);
 }
 
+static inline void ext4_ext_tree_changed(struct inode *inode)
+{
+	EXT4_I(inode)->i_ext_generation++;
+}
+
 static inline void
 ext4_ext_invalidate_cache(struct inode *inode)
 {
Index: linux-2.6.32-el6-beta/fs/ext4/extents.c
===================================================================
--- linux-2.6.32-el6-beta.orig/fs/ext4/extents.c
+++ linux-2.6.32-el6-beta/fs/ext4/extents.c
@@ -1740,6 +1740,7 @@ cleanup:
 		ext4_ext_drop_refs(npath);
 		kfree(npath);
 	}
+	ext4_ext_tree_changed(inode);
 	ext4_ext_invalidate_cache(inode);
 	return err;
 }
@@ -2415,6 +2416,7 @@ static int ext4_ext_remove_space(struct 
 		}
 	}
 out:
+	ext4_ext_tree_changed(inode);
 	ext4_ext_drop_refs(path);
 	kfree(path);
 	ext4_journal_stop(handle);
Index: linux-2.6.32-el6-beta/fs/ext4/ext4.h
===================================================================
--- linux-2.6.32-el6-beta.orig/fs/ext4/ext4.h
+++ linux-2.6.32-el6-beta/fs/ext4/ext4.h
@@ -721,6 +721,8 @@ struct ext4_inode_info {
 	tid_t i_datasync_tid;
 
 	__u64 i_fs_version;
+
+	unsigned long i_ext_generation;
 };
 
 #define HAVE_DISK_INODE_VERSION
