--- linux-stage.orig/fs/ext4/super.c	2012-06-20 05:33:21.000000000 +0000
+++ linux-stage/fs/ext4/super.c	2012-06-20 05:34:51.000000000 +0000
@@ -672,7 +672,6 @@
 	percpu_counter_destroy(&sbi->s_freeinodes_counter);
 	percpu_counter_destroy(&sbi->s_dirs_counter);
 	percpu_counter_destroy(&sbi->s_dirtyblocks_counter);
-	brelse(sbi->s_sbh);
 #ifdef CONFIG_QUOTA
 	for (i = 0; i < MAXQUOTAS; i++)
 		kfree(sbi->s_qf_names[i]);
@@ -686,7 +685,6 @@
 		dump_orphan_list(sb, sbi);
 	J_ASSERT(list_empty(&sbi->s_orphan));
 
-	invalidate_bdev(sb->s_bdev, 0);
 	if (sbi->journal_bdev && sbi->journal_bdev != sb->s_bdev) {
 		/*
 		 * Invalidate the journal device's buffers.  We don't want them
@@ -697,6 +695,8 @@
 	}
 	if (sbi->s_mmp_tsk)
 		kthread_stop(sbi->s_mmp_tsk);
+	brelse(sbi->s_sbh);
+	invalidate_bdev(sb->s_bdev, 0);
 	sb->s_fs_info = NULL;
 	/*
 	 * Now that we are completely done shutting down the
