diff -ur linux-stage.orig/fs/ext4/dynlocks.c linux-stage/fs/ext4/dynlocks.c
--- linux-stage.orig/fs/ext4/dynlocks.c	2011-05-10 12:57:44.000000000 -0400
+++ linux-stage/fs/ext4/dynlocks.c	2011-05-10 13:07:24.000000000 -0400
@@ -16,6 +16,8 @@
 #define DYNLOCK_HANDLE_DEAD	0xd1956ee
 #define DYNLOCK_LIST_MAGIC	0x11ee91e6
 
+typedef struct kmem_cache kmem_cache_t;
+
 static kmem_cache_t * dynlock_cachep = NULL;
 
 struct dynlock_handle {
@@ -38,7 +40,7 @@
 					 sizeof(struct dynlock_handle),
 					 0,
 					 SLAB_HWCACHE_ALIGN,
-					 NULL, NULL);
+					 NULL);
 	if (dynlock_cachep == NULL) {
 		printk(KERN_ERR "Not able to create dynlock cache");
 		rc = -ENOMEM;
