--- linux-stage.orig/fs/ext4/mmp.c	2015-12-03 03:05:49.859815762 +0530
+++ linux-stage/fs/ext4/mmp.c	2015-12-03 03:08:54.342315996 +0530
@@ -99,8 +99,11 @@
 
 	mmp = (struct mmp_struct *)((*bh)->b_data);
 	if (le32_to_cpu(mmp->mmp_magic) != EXT4_MMP_MAGIC ||
-	    !ext4_mmp_csum_verify(sb, mmp))
+	    !ext4_mmp_csum_verify(sb, mmp)) {
+		brelse(*bh);
+		*bh = NULL;
 		return -EINVAL;
+	}
 
 	return 0;
 }
@@ -220,6 +222,7 @@
 					     "The filesystem seems to have been"
 					     " multiply mounted.");
 				ext4_error(sb, "abort");
+				put_bh(bh_check);
 				goto failed;
 			}
 			put_bh(bh_check);
